### PR TITLE
Cereal for Array, ArrayBase

### PR DESCRIPTION
**Pull request by dkeeney**
This adds Array and ArrayBase to Cereal serialization.

### CODE REVIEW

File: src/nupic/engine/Link.cpp [hunk 1]
```diff
@@ -201,7 +201,7 @@ void Link::compute() {
               << " type=" << BasicType::getName(src.getType())
               << " --> " << BasicType::getName(dest.getType()) << std::endl;
 
-	NTA_CHECK(src.getCount() + destOffset_ <= dest.getMaxElementsCount())
+	NTA_CHECK(src.getCount() + destOffset_ <= dest.getCount())
```
Comment (breznak) [inline src/nupic/engine/Link.cpp]: this is unrelated to the PR, so is it actually correct?

File: src/nupic/ntypes/Array.hpp [hunk 2]
```diff
@@ -319,4 +320,9 @@ class Array : public ArrayBase {
 };
 } // namespace nupic
 
+CEREAL_REGISTER_TYPE(nupic::Array)
+CEREAL_REGISTER_POLYMORPHIC_RELATION(nupic::ArrayBase, nupic::Array)
```
Comment (breznak) [inline src/nupic/ntypes/Array.hpp]: this is some advanced stuff, I hope it's only for the Array/Base, which are complicated. 
Maybe add a comment that this is for the complicated pointer stuff, otherwise not needed

File: src/nupic/ntypes/ArrayBase.cpp [hunk 3]
```diff
@@ -95,34 +95,32 @@ ArrayBase::~ArrayBase() {}
...
   if (type_ == NTA_BasicType_SDR) {
     std::vector<UInt> dimension;
     dimension.push_back((UInt)count);
     allocateBuffer(dimension);
   } else {
-    std::shared_ptr<char> sp(new char[capacity_],
+    std::shared_ptr<char> sp(new char[count_ * BasicType::getSize(type_)],
                              std::default_delete<char[]>());
     buffer_ = sp;
-    own_ = true;
```
Comment (breznak) [inline src/nupic/ntypes/ArrayBase.cpp]: nice to see these "owns buffer" go away

File: src/nupic/ntypes/ArrayBase.cpp [hunk 4]
```diff
@@ -190,72 +183,40 @@ const void *ArrayBase::getBuffer() const {
...
-SDR& ArrayBase::getSDR() {
+sdr::SDR& ArrayBase::getSDR() {
   NTA_CHECK(type_ == NTA_BasicType_SDR) << "Does not contain an SDR object";
   if (buffer_ == nullptr) {
     std::vector<UInt> zeroDim;
     zeroDim.push_back(0u);
     allocateBuffer(zeroDim);  // Create an empty SDR object.
   }
-  SDR& sdr = *((SDR *)buffer_.get());
+  sdr::SDR& sdr = *((sdr::SDR *)buffer_.get());
```
Comment (breznak) [inline src/nupic/ntypes/ArrayBase.cpp]: nit, can you use static_cast ? (or other relevant c++ casting), it does better checks that the plain C-cast

File: src/nupic/ntypes/ArrayBase.cpp [hunk 5]
```diff
@@ -190,72 +183,40 @@ const void *ArrayBase::getBuffer() const {
...
 }
-const SDR& ArrayBase::getSDR() const {
+const sdr::SDR& ArrayBase::getSDR() const {
   NTA_CHECK(type_ == NTA_BasicType_SDR) << "Does not contain an SDR object";
   if (buffer_ == nullptr)
     // this is const, cannot create an empty SDR.
-    NTA_THROW << "getSDR: SDR pointer is null";  
-  SDR& sdr = *((SDR *)buffer_.get());
+    NTA_THROW << "getSDR: SDR pointer is null";
+  sdr::SDR& sdr = *((sdr::SDR *)buffer_.get());
```
Comment (breznak) [inline src/nupic/ntypes/ArrayBase.cpp]: ditto, static_cast<*((sdr::SDR *)>(buffer_.get())

File: src/nupic/ntypes/ArrayBase.cpp [hunk 6]
```diff
@@ -291,15 +252,16 @@ bool ArrayBase::has_buffer() const { return (buffer_.get() != nullptr); }
...
     a.allocateBuffer(maxsize);
     a.zeroBuffer();
   }
-  if (offset == 0) {
-    // This could be the first buffer of a Fan-In set.
-    if (a.getCount() != maxsize)
-      a.setCount(maxsize);
-  }
+	// TODO:  Comment this out until we are sure that it is not needed.
+  //if (offset == 0) {
```
Comment (breznak) [inline src/nupic/ntypes/ArrayBase.cpp]: what was this for?Why we don't use it but you wanna keep that?

File: src/nupic/utils/StringUtils.cpp [hunk 7]
```diff
@@ -81,6 +81,24 @@ Real32 StringUtils::toReal32(const std::string& s, bool throwOnError, bool * fai
   return r;
 }
 
+Real64 StringUtils::toReal64(const std::string& s, bool throwOnError, bool * fail) {
```
Comment (breznak) [inline src/nupic/utils/StringUtils.cpp]: I couldn't find where this is used? 
- if unused, rm
- if at only one class, make it private of that class
- otherwise, ok

File: src/test/unit/ntypes/ArrayTest.cpp [hunk 8]
```diff
@@ -31,16 +31,18 @@
...
 #include <memory>
 #include <limits.h>
 
 #include <gtest/gtest.h>
 
 namespace testing {
-    
+
+static bool verbose = false;
+#define VERBOSE if(verbose) std::cerr << "[          ]"
```
Comment (breznak) [inline src/test/unit/ntypes/ArrayTest.cpp]: OT: can we make some meta-include, so this is included in all tests? Maybe even useful for non-test classes, add to Log.hpp? (consider in a separate Issue)

File: src/nupic/engine/Link.cpp (hunk 1 above)
Comment (breznak) [reply src/nupic/engine/Link.cpp]: ..and it's intended, maxElementCount removed

File: src/nupic/ntypes/Array.hpp (hunk 2 above)
Comment (dkeeney) [reply src/nupic/ntypes/Array.hpp]: Any time you have a base class and an instance class and if you serialize the instance and some of the fields you want serialized happen to be in the base class, then you apparently need to do this.  So far Array and ArrayBase are the only place that this seems to be the case.

File: src/nupic/ntypes/ArrayBase.cpp (hunk 3 above)
Comment (dkeeney) [reply src/nupic/ntypes/ArrayBase.cpp]: > nice to see these "owns buffer" go away

That was for when Array was also handling external buffers and Sparse data which had variable length.  I am just now finally getting around to removing some of the leftover stuff.

File: src/nupic/ntypes/ArrayBase.cpp (hunk 4 above)
Comment (dkeeney) [reply src/nupic/ntypes/ArrayBase.cpp]: > nit, can you use static_cast ?

I just ever got into the habit of using static_cast.  I know what C casting does.  Not so sure about the C++ casting.  But I guess I should lean to use the new stuff.

File: src/nupic/ntypes/ArrayBase.cpp (hunk 6 above)
Comment (dkeeney) [reply src/nupic/ntypes/ArrayBase.cpp]: It was for handling Fan-in merge when buffers could be variable length (Sparse arrays).  I don't think any other code was depending on this logic being here so if nothing breaks I will remove the commented out code.

File: src/nupic/utils/StringUtils.cpp (hunk 7 above)
Comment (dkeeney) [reply src/nupic/utils/StringUtils.cpp]: > I couldn't find where this is used?

It isn't.  I added it because it was the only one of the data types that was not included.  It was in the .hpp but it had not been implemented.   It is very likely that none of the toXXX functions are used.

File: src/test/unit/ntypes/ArrayTest.cpp (hunk 8 above)
Comment (dkeeney) [reply src/test/unit/ntypes/ArrayTest.cpp]: > static bool verbose = false;
#define VERBOSE if(verbose) std::cerr << "[          ]"
can we make some meta-include, so this is included in all tests? 

Well, the `[             ]` part is intended to match up with the gtest output so that your text does not mess up the `[     OK]` column.  I don't think this applies much outside of gtest.  You would not want the `static bool verbose` to be global.  

So, lets not make this common.  We can add these two lines if we want to use this VERBOSE macro in a test.

File: src/nupic/ntypes/Array.hpp (hunk 2 above)
Comment (breznak) [reply src/nupic/ntypes/Array.hpp]: > So far Array and ArrayBase are the only place that this seems to be the case. 

Encoders & BaseEncoder too. But that's not that bad once you know about it :+1:

File: src/nupic/ntypes/ArrayBase.cpp (hunk 4 above)
Comment (breznak) [reply src/nupic/ntypes/ArrayBase.cpp]: have a look, it's a nice c++ feature: https://stackoverflow.com/questions/332030/when-should-static-cast-dynamic-cast-const-cast-and-reinterpret-cast-be-used 

My layman's version: 
- it's possible to grep sourcecode for the cast
- no performance penalty
- does some checks
- 1st goto version: `static_cast<T>(val)`

File: src/nupic/utils/StringUtils.cpp (hunk 7 above)
Comment (breznak) [reply src/nupic/utils/StringUtils.cpp]: > It is very likely that none of the toXXX functions are used. 

let's get killing LOC!

**Comment by dkeeney** [comment]
I will need another review before the master changes again.  Thanks.

**Comment by breznak** [comment]
> I will need another review before the master changes again. Thanks. 

I will, but you haven't made changes on the last review (?)

File: src/nupic/ntypes/ArrayBase.cpp (hunk 4 above)
Comment (breznak) [reply src/nupic/ntypes/ArrayBase.cpp]: ..I'd like to see these implemented for the PR

**Comment by dkeeney** [comment]
> I will, but you haven't made changes on the last review (?)

So, you really are going to make me change my coding style. 
Your are right you know....but that does not make it any easier.

**Comment by dkeeney** [comment]
Hmmm,
When I compile now after pulling in the master, I am getting the following:
```
2>c:\projects\ai\cppb\src\nupic\algorithms\temporalmemory.cpp(141): warning C4267: 'argument': conversion from 'size_t' to 'nupic::algorithms::connections::CellIdx', possible loss of data
2>c:\projects\ai\cppb\src\nupic\algorithms\temporalmemory.cpp(160): warning C4267: 'initializing': conversion from 'size_t' to 'nupic::UInt32', possible loss of data
2>c:\projects\ai\cppb\src\nupic\algorithms\temporalmemory.cpp(160): warning C4267: 'initializing': conversion from 'size_t' to 'const nupic::UInt32', possible loss of data
2>c:\projects\ai\cppb\src\nupic\algorithms\temporalmemory.cpp(286): warning C4267: 'argument': conversion from 'size_t' to 'nupic::Int', possible loss of data
2>c:\projects\ai\cppb\src\nupic\algorithms\temporalmemory.cpp(571): warning C4267: 'argument': conversion from 'size_t' to '_Ty', possible loss of data
2>c:\projects\ai\cppb\src\nupic\algorithms\temporalmemory.cpp(575): warning C4267: 'argument': conversion from 'size_t' to '_Ty', possible loss of data
2>c:\projects\ai\cppb\src\nupic\algorithms\temporalmemory.cpp(585): warning C4267: 'initializing': conversion from 'size_t' to 'nupic::UInt32', possible loss of data
2>c:\projects\ai\cppb\src\nupic\algorithms\temporalmemory.cpp(585): warning C4267: 'initializing': conversion from 'size_t' to 'const nupic::UInt32', possible loss of data

and lots of them
```

I am going to assume that is not me.

**Comment by dkeeney** [comment]
I am going to be away from my computer for the rest of the day.
Will have the casts fixed shortly.

File: src/test/unit/ntypes/ArrayTest.cpp [hunk 9]
```diff
@@ -39,7 +39,7 @@
 
 namespace testing {
 
-static bool verbose = false;
+static bool verbose = true;
```
Comment (breznak) [inline src/test/unit/ntypes/ArrayTest.cpp]: TODO probably debugging this now, later toggle to silent

File: src/nupic/utils/StringUtils.cpp (hunk 7 above)
Comment (breznak) [reply src/nupic/utils/StringUtils.cpp]: moved to #378

**Comment by dkeeney** [comment]
I have lots more cast fixes coming...so don't push to master yet.

**Comment by breznak** [comment]
> I have lots more cast fixes coming...so don't push to master yet.

oh, you..stop it, you are spoiling me with goodness :smile:

**Comment by dkeeney** [comment]
I am still getting errors caused by the type of CellIdx
```
2>TemporalMemory.cpp
2>c:\projects\ai\cppb\src\nupic\algorithms\temporalmemory.cpp(141): warning C4267: 'argument': conversion from 'size_t' to 'nupic::algorithms::connections::CellIdx', possible loss of data
2>c:\projects\ai\cppb\src\nupic\algorithms\temporalmemory.cpp(160): warning C4267: 'initializing': conversion from 'size_t' to 'nupic::UInt32', possible loss of data
2>c:\projects\ai\cppb\src\nupic\algorithms\temporalmemory.cpp(160): warning C4267: 'initializing': conversion from 'size_t' to 'const nupic::UInt32', possible loss of data
2>c:\projects\ai\cppb\src\nupic\algorithms\temporalmemory.cpp(286): warning C4267: 'argument': conversion from 'size_t' to 'nupic::Int', possible loss of data
2>c:\projects\ai\cppb\src\nupic\algorithms\temporalmemory.cpp(571): warning C4267: 'argument': conversion from 'size_t' to '_Ty', possible loss of data
2>        with
2>        [
2>            _Ty=std::seed_seq::result_type
2>        ]
2>c:\projects\ai\cppb\src\nupic\algorithms\temporalmemory.cpp(575): warning C4267: 'argument': conversion from 'size_t' to '_Ty', possible loss of data
2>        with
2>        [
2>            _Ty=std::seed_seq::result_type
2>        ]
2>c:\projects\ai\cppb\src\nupic\algorithms\temporalmemory.cpp(585): warning C4267: 'initializing': conversion from 'size_t' to 'nupic::UInt32', possible loss of data
2>c:\projects\ai\cppb\src\nupic\algorithms\temporalmemory.cpp(585): warning C4267: 'initializing': conversion from 'size_t' to 'const nupic::UInt32', possible loss of data
```

**Comment by dkeeney** [comment]
Does your compiler not show those warnings?

**Comment by dkeeney** [comment]
And....
```
7>TemporalMemoryTest.cpp
7>c:\projects\ai\cppb\src\test\unit\algorithms\temporalmemorytest.cpp(1360): warning C4267: 'initializing': conversion from 'size_t' to 'int', possible loss of data
7>c:\projects\ai\cppb\src\test\unit\algorithms\temporalmemorytest.cpp(1368): warning C4267: 'initializing': conversion from 'size_t' to 'nupic::Int', possible loss of data
7>c:\projects\ai\cppb\src\test\unit\algorithms\temporalmemorytest.cpp(1560): warning C4267: '/=': conversion from 'size_t' to '_Ty', possible loss of data
7>        with
7>        [
7>            _Ty=std::seed_seq::result_type
7>        ]
```

**Comment by dkeeney** [comment]
To be consistent....we should make ALL size related things size_t.  This includes the dimensions.

**Comment by dkeeney** [comment]
Oh, and what about the sparse array elements... aren't they also size related?   Probably should keep those at UInt32.

**Comment by breznak** [comment]
> I am still getting errors caused by the type of CellIdx

No! (gcc 7.3, linux), I thought those had gone away for you. Try a clean build... 

> Oh, and what about the sparse array elements... aren't they also size related? Probably should keep those at UInt32. 

we can't end up with everything size_t'ed :) SDR_sparse_t elements (are now sdr::ElemSparse type) and are/must be matched with connections::CellIdx (so yes, now UInt32, but can change)

**Comment by dkeeney** [comment]
> I am still getting errors caused by the type of CellIdx

I see them only when I do a clean build.

**Comment by breznak** [comment]
> I see them [CellIdx typedef related errors] only when I do a clean build. 

That's very strange! can you open an issue for that? Event the CI guys here are satisfied :thinking:

**Comment by dkeeney** [comment]
I am running a unit test and if that passes I will push some more changes.

**Comment by dkeeney** [comment]
crashed.
Guess I have more work to do.

**Comment by breznak** [comment]
> crashed.
Guess I have more work to do.
 
let's merge this, and the other can come later..

**Comment by dkeeney** [comment]
Maybe you can help with this....
```
  template<class Archive>
  void load_ar(Archive & ar) {
    std::vector<UInt>& v = *(reinterpret_cast<std::vector<UInt>*>(this));
    ar(v);
  }
```
Trying to serialize the Dimensions class.  This is a subclass of std::vector<UInt>
What I thought I could do was use the this pointer to tell Cereal that this is really a vector.

The above code crashes deep inside the cereal templates (about 5 levels deep).
Access violation reading location 0xFFFFFFFFFFFFFF in other words it was lost.

So obviously it does not like something.

I have a social engagement in 15 min so I will take this up in the morning.  Will push what I have so you can review all of casts..... you asked for them.

**Comment by breznak** [comment]
> can review all of casts..... you asked for them. 

I've reverted the last 2 changes in this PR, to merge this as "cereal", and I'll put them in a new one. 

> std::vector<UInt>& v = *(reinterpret_cast<std::vector<UInt>*>(this)); 

maybe a static/dynamic cast? 

PS: the new extra casts break some tests, we can fix them, but no need to sweat it. I meant please use the casts for new code, but we don't have to convert all the existing codebase atonce now.

**Comment by dkeeney** [comment]
> I've reverted the last 2 changes in this PR, to merge this as "cereal", and I'll put them in a new one.

Ok, I will continue with the Cereal conversions.  I have one more PR for Spec and friends (Collection, Scaler, and Value).   That would complete all of the lower level modules in ntypes and types.  Then I can start in on algorithms, followed by regions and then NetworkAPI.
